src/Component.ts

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) 2015 NAVER Corp.
+ * egjs projects are licensed under the MIT license
+ */
+
+function isUndefined(value: any): boolean {
+    return typeof value === "undefined";
+}
+
+interface DefaultParam<T extends EventMap> {
+  eventType: string;
+  stop: () => void;
+  currentTarget: Component<T>;
```

Comment:
@WoodNeck 
If inherited, currentTarget is dangerous. Use this type.

Reply:
Well it actually interprets `this` as `DefaultParam`, so I can't use it.

Reply:
@WoodNeck 

If you use currentTarget you can't use the method?

Reply:
Yes, in both cases.

Reply:
@WoodNeck 

Then to use currentTarget, you would have to use any ?

If so, try putting `this` into generic in class, not interface.
```
interface DefaultParam<T> {
  currentTarget: T;
}
```

Reply:
`this` is not applicable for generic types, can you show me an actual example on how to use that?

Reply:
```ts
type EventCallback<T extends EventMap, K extends EventKey<T>, F>
  = T[K] extends NoArguments
    ? (event: DefaultProps<F>) => any
    : T[K] extends (evt: infer U, ...restParam: infer V) => any
      ? (event: U & DefaultProps<F>, ...restParam: V) => any
      : (event: T[K] & DefaultProps<F>) => any;
type EventHash<T extends EventMap, F> = Partial<{ [K in EventKey<T>]: EventCallback<T, K, F> }>;

public on(eventHash: EventHash<T, this>): this;
```

Reply:
That works!